### DISCUSSION TITLE
Biome formatter without auto-wrapping

### DISCUSSION THREAD

**Discussion by fregante** [biomejs/biome#3493] [Q&A] [answered]
I don't use Prettier in a specific project, but I do want some formatting rules (brake spacing, etc), for which I've been using ESLint.

Is it possible to use Biome as a formatter (JS) without enforcing a specific line width? I tried using `"lineWidth": 180` but that means every array is still changed to a single line. I just want Biome to _not_ touch the wrapping unless there's a specific reason to (e.g. _don't allow two instructions on a line: `a(); b()`_)

**Comment by KalebKloppe** [comment]
I agree, this is why I am looking for a new formatter. I need a formatter that lets the writer decide where to break the code. Prettier [closed the discussion](https://github.com/prettier/prettier/issues/4298) on this. 

Syntax-highlighted code isn't read like paragraphs in a book, from left to right, top to bottom. People jump around looking for keywords and symbols. Sometimes the keywords and symbols are predictable and can be hidden off the page, but other times you might want to call attention to them.

I wish I could break the format intentionally:

```ts
const boilerplateStuff = ["dont", "worry", "about", "these"]
const moreBoilerplate = ["more", "stuff", "you", "don't", "need", "to", "focus", "on"]
const extraBoilerplateStuff = ["extra", "boilerplate", "that", "is", "not", "important"]
const importantStuff = [
  "gee",
  "these",
  "seem",
  "important"
]
```

Other times, I want to apply a consistent format, but the line length enforcement doesn't allow:
```ts
{ orderId: 1, customerName: "John Doe", orderTotal: 99.99, status: "Delivered" },
{ orderId: 2, customerName: "Jane Smith", orderTotal: 99.99, status: "Delivered" },
{ 
    orderId: 3, 
    customerName: "AnUnfortunately LongName", 
    orderTotal: 99.99, 
    status: "Delivered" 
},
{ orderId: 4, customerName: "Bob Brown", orderTotal: 99.99, status: "Delivered" },
```

**Reply by Conaclos** [reply, marked as answer]
Have you tried [dprint](https://dprint.dev/)? I think it could fulfill your needs.

Personally, I like the idea of a way to turn off auto-wrapping. However, this could be a significant change to the way Biome works. Also, it is not a high priority change, as it is already a lot of work to adjust Prettier.

**Reply by olejorgenb** [reply]
> Personally, I like the idea of a way to turn off auto-wrapping

Then perhaps it deserve an issue to better track support? Instead of being buried in an "answered" discussion.

**Comment by ematipico** [comment]
FYI, we have a page that talks about the Biome formatter option philosophy: https://biomejs.dev/formatter/option-philosophy/

**Reply by KalebKloppe** [reply]
In that page you mention

> the chance to start fresh and avoid the pitfalls that Prettier had fallen into with its other existing options

It's important to note that the typographic advice for 'ideal line width' originates from formatting _body text_ of a book, not a series of React components. Even in the context of body text, the ideal is typically presented as a range.

I suggest considering a `minLineWidth` and `maxLineWidth` (like eslint used to have) instead of a single `lineWidth` value. 
For backwards compatibility, you could set the `lineWidth` value as both the minimum and maximum. The behavior would remain the same.

**Reply by firefoxic** [reply]
You have done an amazing job and created a wonderful tool. But the mere fact that it is impossible to disable this harmful setting makes it worthless to me 😒

**Comment by RafaelKr** [comment]
It would be great to have the `consistent` rule from eslint available. So the developer can decide if the array should be wrapped or not. But if the developer wraps it, it must be consistently wrapped.

Please see the examples here: https://eslint.org/docs/latest/rules/array-element-newline#consistent

There are a lot of existing discussions around this issue in the prettier ecosystem:
- https://github.com/prettier/prettier/issues/2716 (63 comments)
- https://github.com/prettier/prettier-vscode/issues/352 (42 comments)
- https://github.com/prettier/prettier/issues/8507 (general issue listing some others)
- https://github.com/prettier/prettier/issues/4424#issuecomment-1301504204 (47 comments, highest voted comment is related)

Edit: When the developer decides to wrap every array item on an own line usually this is on purpose and Biome shouldn't revert this decision of the developer. Wrapping each item on an own line improves readability, quicker reordering of items and when working in a team it leads to less complex merge conflicts, when multiple developers change the same array.

**Comment by AlekseyMelikov** [comment]
A very useful and necessary feature

**Comment by olejorgenb** [comment]
To me, the lack of any explicit control of one-item-per-line is BIG drawback.

The python formatters `ruff` and `black` have this "magic trailing comma" rule to handle this situation which I think is nice:

> A pre-existing trailing comma informs Black to always explode contents of the current bracket pair into one item per line. Read more about this in the [Pragmatism](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#pragmatism) section below.

The only argument I can see against having such a rule is that changing the max width would not cause rewrapping. (since we want to multi-line list to always have a trailing comma). But as long as the rule can be turned off to force rewrapping in this rare case I don't see it as much of an issue.

Personally I wouldn't mind just using the https://prettier.io/docs/options#object-wrap convention everywhere either. Yes, if I shorten a wrapped list I would need to remove that line break to "one-line", but is it really so bad? And is it really so bad if some list are "inconsistently" wrapped?

**Comment by tlaurie-cl** [comment]
This is a dealbreaker that's convinced us to stay with Prettier, despite the slower performance.

@KalebKloppe's [second example](https://github.com/biomejs/biome/discussions/3493#discussioncomment-10120798) is a good illustration of the problem. But here's another:

Given this Nx json target configuration:
```json
"some-nx-target": {
  "dependsOn": [
    "some-other-target-a",
    "some-other-target-b",
    "foo",
    "bar",
    "baz"
  ]
}
```

Biome will format it as:

```json
"some-nx-target": {
  "dependsOn": ["some-other-target-a", "some-other-target-a", "foo", "bar", "baz"]
}
```

Many medium-to-long arrays like this one have similarities between items that can be leveraged to detect-errors, multiline-edit, and isolate diffs. Arbitrarily forcing items onto one line makes all of that more difficult.

The same wrapping behavior in Prettier prevented us from using it for a long time until finding the [prettier-plugin-multiline-arrays](https://www.npmjs.com/package/prettier-plugin-multiline-arrays) plugin. There is no equivalent in Biome.

I think Prettier just got this wrong originally and that using `printWidth` (`lineWidth`) as the heuristic for array wrapping was an underdeveloped idea. 

[prettier-plugin-multiline-arrays](https://www.npmjs.com/package/prettier-plugin-multiline-arrays) is _the_ most popular behavior-changing plugin for Prettier and has [half the monthly downloads as dprint](https://npmtrends.com/dprint-vs-prettier-plugin-multiline-arrays). There is demand, its just a half-step past Prettier-parity.

**Reply by tlaurie-cl** [reply]
Maybe its best just to wait for [oxfmt](https://github.com/oxc-project/oxc/discussions/13608#discussioncomment-14348599)'s release.

**Reply by ematipico** [reply]
This discussion is very old. Biome now has an option that fits your needs: https://biomejs.dev/reference/configuration/#formatterexpand

**Reply by PAStheLoD** [reply]
Unfortunately `auto` doesn't work with arrays. (Or maybe I am holding it wrong?)

And it would be amazing to have a "don't touch" option too (for arrays & objects) for certain files/sections.

Also maybe an option to turn off formatting related taṇhā? ^^

**Reply by ematipico** [reply]
Is it an array literal? If so, then it's a bug

**Reply by tlaurie-cl** [reply]
Just tested again, making sure that `"expand": "auto"` was explicitly set, and got the same undesired results.

According to the [docs](https://biomejs.dev/reference/configuration/#formatterexpand):

> object literals are formatted on multiple lines if the first property has a newline, and array literals are formatted on a single line if it fits in the line.

and the [source code](https://github.com/biomejs/biome/blob/5fc24f4c67a6de9105d460a59c8f613d14d58c4d/crates/biome_formatter/src/lib.rs#L799):

> Objects are expanded when the first property has a leading newline. Arrays are always expanded if they are shorter than the line width.

`"expand": "auto"` only affects objects. Arrays can only wrap  `"never"` or `"always"` depending on `lineWidth`.

It seems like this option is supposed to match Prettier's [objectWrap](https://prettier.io/docs/options#object-wrap) option, which also doesn't affect arrays.

**Reply by ematipico** [reply]
Can you file an issue with a reproduction?

**Reply by tlaurie-cl** [reply]
Done 👍
https://github.com/biomejs/biome/issues/8505
